libraries/Microsoft.Bot.Builder.Azure/CosmosDbStorage.cs

```diff
@@ -0,0 +1,267 @@
+        /// More information at https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.connectionpolicy?view=azure-dotnet
+        /// </remarks>
+        /// <param name="serviceEndpoint">The endpoint Uri for the service endpoint from the Azure Cosmos DB service.</param>
+        /// <param name="authKey">The AuthKey used by the client from the Azure Cosmos DB service.</param>
+        /// <param name="databaseId">The Database ID.</param>
+        /// <param name="collectionId">The Collection ID.</param>
+        /// <param name="connectionPolicyConfigurator">
+        ///     A Delegate to further customize the ConnectionPolicy to CosmosDB.
+        ///     More information at https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.connectionpolicy?view=azure-dotnet
+        /// </param>
+        public CosmosDbStorage(Uri serviceEndpoint, string authKey, string databaseId, string collectionId, Action<ConnectionPolicy> connectionPolicyConfigurator = null)
+        {
+            if (serviceEndpoint == null)
+            {
+                throw new ArgumentNullException(nameof(serviceEndpoint), "Service EndPoint for CosmosDB is required.");
+            }
+
+            if (string.IsNullOrEmpty(authKey))
+            {
+                throw new ArgumentException("AuthKey for CosmosDB is required.", nameof(authKey));
+            }
+
+            if (string.IsNullOrEmpty(databaseId))
+            {
+                throw new ArgumentException("DatabaseId is required.", nameof(databaseId));
+            }
+
+            if (string.IsNullOrEmpty(collectionId))
+            {
+                throw new ArgumentException("CollectionId is required.", nameof(collectionId));
+            }
+
+            _databaseId = databaseId;
+            _collectionId = collectionId;
+
+            // Inject BotBuilder version to CosmosDB Requests
+            var version = GetType().Assembly.GetName().Version;
+            var connectionPolicy = new ConnectionPolicy()
+            {
+                UserAgentSuffix = $"Microsoft-BotFramework {version}"
+            };
+
+            // Invoke CollectionPolicy delegate to further customize settings
+            connectionPolicyConfigurator?.Invoke(connectionPolicy);
+            _client = new DocumentClient(serviceEndpoint, authKey, connectionPolicy);
+
+            // Delayed Database and Collection creation if they do not exist. This is a thread-safe operation.
+            _collectionReady = new Lazy<string>(() =>
+            {
+                _client.CreateDatabaseIfNotExistsAsync(new Database { Id = _databaseId })
+                    .Wait();
+
+                var task = _client.CreateDocumentCollectionIfNotExistsAsync(UriFactory.CreateDatabaseUri(_databaseId), new DocumentCollection { Id = _collectionId });
+                task.Wait();
+
+                return task.Result.Resource.SelfLink;
+            });
```

Comment:
I've advised against this in other PRs, but I don't know if everyone's read those, so let me say again here: I would suggest not using this pattern of a `Lazy<T>` when it comes to guarding against something that needs to use `async` operations. 

I suggest creating an async factory method instead, so maybe  `public async ValueTask<Uri> GetCollectionLink()` which essentially wraps a private field that holds that `Uri` and, if it's not initialized yet, you go off and run the `async` code to go get it, otherwise you return the value straightaway. Then, instead of accessing `Lazy<>.Value` everywhere, you just `await` the method instead. 

Now, one of the things you might have be relying on the `Lazy<>` for is serializing the execution/publication so it only happens once. I would argue this is not required in this case because if two threads race to generate the value, it's non-destructive because we're using `XXXIfNotExists` methods and the worst case is one thread made two extra calls to Azure. If that's truly a concern, just add a `lock()` around a gate object instance... it just adds a tad bit more complexity that might not really be worth it.

Finally, I suggest using a `ValueTask<>` vs `Task<>` is to reduce the GC overhead of all the calls after the "first" call.

Reply:
Hi @drub0y ,
How about:

````C#
private string _collectionLink = null;

/// <summary>
/// Delayed Database and Collection creation if they do not exist.
/// </summary>
private async ValueTask<string> GetCollectionLink()
{
    if(_collectionLink == null)
    {
        await _client.CreateDatabaseIfNotExistsAsync(new Database { Id = _databaseId }).ConfigureAwait(false);

        var response = await _client.CreateDocumentCollectionIfNotExistsAsync(UriFactory.CreateDatabaseUri(_databaseId), new DocumentCollection { Id = _collectionId }).ConfigureAwait(false);
        _collectionLink = response.Resource.SelfLink;
    }

    return _collectionLink;
}
````

Usage:
```C#
var collectionLink = await GetCollectionLink();
```